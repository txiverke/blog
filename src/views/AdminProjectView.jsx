// @flow

import React from 'react'
import Helmet from 'react-helmet'
import { connect } from 'react-redux'
import ReactMessages from 'react-messages'
import { Redirect } from 'react-router-dom'

import Loader from '../components/Loader'
import AdminProjectItem from '../components/AdminProjectItem'
import { loadProjectItem, updateProjectData } from '../actions/projectActionCreators'
import { isAuthenticated } from '../actions/userActionCreators'

type Props = {
  dispatch: Function,
  item: Item,
  authenticate: Auth,
  match: Object
}

type State = {
  data: Object
}

class AdminProjectView extends React.Component<Props, State> {
  state: State = {
    data: {}
  }

  props: Props

  componentDidMount() {
    const { dispatch, match } = this.props
    dispatch(loadProjectItem(match.params.id))
    dispatch(isAuthenticated())
  }

  componentWillReceiveProps(nextProps: Object) {
    this.setState({ data: nextProps.item.data})
  }

  updateProject = (obj: Object, id: string) => {
    if(Object.keys(obj).length && id.length) {
      const { dispatch } = this.props
      dispatch(updateProjectData(obj, id))
    }
  }

  render() {
    const { message, error, completed } = this.props.item
    const { data } = this.state
    const isAuthenticate = this.props.authenticate.data

    if (!isAuthenticate) return <Redirect to="/" />

    if (isAuthenticate && completed) {
      return (
        <section className="app-view pb5">
          <Helmet 
            title="Admin page" 
            meta={[
              { name: "description", content: "Admin page" },
              { property: "og:title", content: "Admin page" }
            ]}
          />  
          <ReactMessages message={message} error={error} next={true} />
          <AdminProjectItem 
            label="Update project" 
            handleProject={(obj, id) => this.updateProject(obj, id)} 
            data={data}
          />
        </section>
      )
    } 

    return (
      <div className="app-view app-view-centered">
        <Loader msg={message} />
      </div>
    )
  }
}

const mapStateToProps = state => ({ 
  authenticate: state.authenticate,
  item: state.item, 
})

export const Unwrapped = AdminProjectView
export default connect(mapStateToProps)(AdminProjectView)