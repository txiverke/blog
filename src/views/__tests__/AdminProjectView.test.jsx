// @flow

import React from 'react'
import { shallow, mount } from 'enzyme'
import { Unwrapped as UnwrappedAdminProjectView } from '../AdminProjectView'

function setup() {
  const props = {
    dispatch: jest.fn(),
    item: {
      data: {},
      message: 'Just a fake message',
      error: false,
      completed: true
    },
    authenticate: { 
      completed: true,
      data: '',
      message: '',
      error: false
    },
    match: { params: { id: '1'}}
  }

  const enzymeWrapper = mount(<UnwrappedAdminProjectView {...props} />)

  return {
    props,
    enzymeWrapper
  }
}

describe('<AdminProjectView />', () => {
  it('should render itself', () => {
    const dispatch = jest.fn()
    const item = {
      data: {},
      message: 'Just a fake message',
      error: false,
      completed: true
    }
    const authenticate = { 
      completed: true,
      data: '',
      message: '',
      error: false
    }
    const match = { params: { id: '1'}}
    const component = shallow(<UnwrappedAdminProjectView 
      dispatch={dispatch}
      item={item}
      authenticate={authenticate}
      match={match}
    />)

    expect(component).toMatchSnapshot()
  })

  it('should render itself', () => {
    const dispatch = jest.fn()
    const item = {
      data: {},
      message: 'Just a fake message',
      error: false,
      completed: false
    }
    const authenticate = { 
      completed: true,
      data: '',
      message: '',
      error: false
    }
    const match = { params: { id: '1'}}
    const component = shallow(<UnwrappedAdminProjectView 
      dispatch={dispatch}
      item={item}
      authenticate={authenticate}
      match={match}
    />)

    expect(component.find('Loader').length).toEqual(1)
  })

})