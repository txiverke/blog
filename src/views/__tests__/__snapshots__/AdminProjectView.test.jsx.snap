// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminProjectView /> should render itself 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminProjectView
    authenticate={
        Object {
            "completed": true,
            "data": "",
            "error": false,
            "message": "",
          }
    }
    dispatch={[Function]}
    item={
        Object {
            "completed": true,
            "data": Object {},
            "error": false,
            "message": "Just a fake message",
          }
    }
    match={
        Object {
            "params": Object {
              "id": "1",
            },
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Loader
        msg="Just a fake message"
/>,
      "className": "app-view app-view-centered",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "msg": "Just a fake message",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Loader
          msg="Just a fake message"
/>,
        "className": "app-view app-view-centered",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "msg": "Just a fake message",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
